api/src/main/java/com/cloud/storage/Snapshot.java

```diff
@@ -48,7 +48,7 @@ public boolean equals(String snapshotType) {
     }
 
     public enum State {
-        Allocated, Creating, CreatedOnPrimary, BackingUp, BackedUp, Copying, Destroying, Destroyed,
+        Allocated, AllocatedVM, Creating, CreatingForVM, CreatedOnPrimary, CreatedOnPrimaryForVM, BackingUp, BackingUpForVM, BackedUp, BackedUpForVM, Copying, Destroying, Destroyed,
```

Comment:
@slavkap Some questions.
- Does the snapshot operation here triggered through VM snapshot operation (as there are changes in VM snapshot strategies)?
- Are the actual volume snapshots taken (for the volumes of the VM) are maintained at volume level (in snapshots table)?
- Does the snapshots taken here are listed with VM snapshots or volume snapshots?

If the snapshots here are VM level snapshots, i think it is better to move these new states (and associated state transitions) to  VM Snapshot object.

Reply:
> * Does the snapshot operation here triggered through VM snapshot operation (as there are changes in VM snapshot strategies)?

Yes, it's triggered through the VM snapshot operation.
> 
> * Are the actual volume snapshots taken (for the volumes of the VM) are maintained at volume level (in snapshots table)?
> 

They are maintained mostly at the VM level. You will see a volume snapshot, but you don't have any options to operate separately on it.

> * Does the snapshots taken here are listed with VM snapshots or volume snapshots?

You'll see both a VM snapshot and also a volume snapshot for each volume of the group snapshot.

![image](https://user-images.githubusercontent.com/51903378/128169186-078024b6-19f5-48ba-b1dc-5988f9019057.png)
![image](https://user-images.githubusercontent.com/51903378/128169238-460e0afe-b01f-48ee-bfb5-2ceb6469a131.png)

Reply:
> > * Does the snapshot operation here triggered through VM snapshot operation (as there are changes in VM snapshot strategies)?
> 
> Yes, it's triggered through the VM snapshot operation.
> 
> > * Are the actual volume snapshots taken (for the volumes of the VM) are maintained at volume level (in snapshots table)?
> 
> They are maintained mostly at the VM level. You will see a volume snapshot, but you don't have any options to operate separately on it.
> 
> > * Does the snapshots taken here are listed with VM snapshots or volume snapshots?
> 
> You'll see both a VM snapshot and also a volume snapshot for each volume of the group snapshot.
> 
> ![image](https://user-images.githubusercontent.com/51903378/128169186-078024b6-19f5-48ba-b1dc-5988f9019057.png)
> ![image](https://user-images.githubusercontent.com/51903378/128169238-460e0afe-b01f-48ee-bfb5-2ceb6469a131.png)

@slavkap As no operations can be performed on the volume snapshots (of the VM) and these are read-only, I feel it is better to keep the snapshot as VM snapshot only and maintain the relation of these volume snapshots to VM snapshot at the backend (in DB). Also, better to move these new states (and associated state transitions) to VM Snapshot object.